In [1]:
import requests
import os
from pprint import pprint
import pandas as pd
from transformers import pipeline
from typing import List, Dict

In [2]:
with open('../../../../api_keys/nytimes/api_key.txt') as f:
    apikey = f.readline()

In [3]:
# # Only returns 10 ata  time (need to explore more)

# # Article Search: https://developer.nytimes.com/docs/articlesearch-product/1/routes/articlesearch.json/get to explore API
# # https://api.nytimes.com/svc/search/v2/articlesearch.json?q=<QUERY>&api-key=<APIKEY>

# query = "politics"
# begin_date = "20200101"  # YYYYMMDD
# filter_query = "\"body:(\"Trump\") AND glocations:(\"WASHINGTON\")\""  # http://www.lucenetutorial.com/lucene-query-syntax.html
# page = "0"  # <0-100>
# sort = "relevance"  # newest, oldest
# query_url = f"https://api.nytimes.com/svc/search/v2/articlesearch.json?" \
#             f"q={query}" \
#             f"&api-key={apikey}" \
#             f"&begin_date={begin_date}" \
#             # f"&fq={filter_query}" \
#             # f"&page={page}" \
#             # f"&sort={sort}"

# r = requests.get(query_url)
# # pprint(r.json())

In [4]:
year = "2020"  # <1851 - 2020>
month = "7"  # <1 - 12>
query_url = f"https://api.nytimes.com/svc/archive/v1/{year}/{month}.json?api-key={apikey}"

r = requests.get(query_url)
# pprint(r.json())

In [5]:
data = pd.json_normalize(r.json()['response']['docs'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6553 entries, 0 to 6552
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   abstract                 6553 non-null   object
 1   web_url                  6553 non-null   object
 2   snippet                  6553 non-null   object
 3   lead_paragraph           6553 non-null   object
 4   print_section            4078 non-null   object
 5   print_page               4078 non-null   object
 6   source                   6553 non-null   object
 7   multimedia               6553 non-null   object
 8   keywords                 6553 non-null   object
 9   pub_date                 6553 non-null   object
 10  document_type            6553 non-null   object
 11  news_desk                6553 non-null   object
 12  section_name             6553 non-null   object
 13  subsection_name          2646 non-null   object
 14  type_of_material         6445 non-null  

In [6]:
data.head()

abstract  \
0  A small-time businessman became a key middlema...   
1  The so-called Capitol Hill Organized Protest a...   
2  Surging outbreaks in the U.S. Embassy and the ...   
3  Journalists have been wary of Alden Global Cap...   
4  Adam Hollingsworth, known in Chicago as the Dr...   

                                             web_url  \
0  https://www.nytimes.com/2020/07/01/world/asia/...   
1  https://www.nytimes.com/2020/07/01/us/seattle-...   
2  https://www.nytimes.com/2020/07/01/us/politics...   
3  https://www.nytimes.com/2020/07/02/business/me...   
4  https://www.nytimes.com/2020/07/01/style/dread...   

                                             snippet  \
0  A small-time businessman became a key middlema...   
1  The so-called Capitol Hill Organized Protest a...   
2  Surging outbreaks in the U.S. Embassy and the ...   
3  Journalists have been wary of Alden Global Cap...   
4  Adam Hollingsworth, known in Chicago as the Dr...   

                                      lead_paragraph print_section print_page  \
0  KABUL, Afghanistan — He was a lowly drug smugg...             A         18   
1  SEATTLE — For weeks, officials in Seattle have...             A         20   
2  WASHINGTON — Inside the sprawling American Emb...             A          1   
3  Alden Global Capital seemed in position this w...             B          1   
4  In late May, as protests against police brutal...            ST          2   

               source                                         multimedia  \
0  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
1  The New York Times                                                 []   
2  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
3  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
4  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   

                                            keywords  \
0  [{'name': 'glocations', 'value': 'Russia', 'ra...   
1  [{'name': 'subject', 'value': 'George Floyd Pr...   
2  [{'name': 'subject', 'value': 'Coronavirus (20...   
3  [{'name': 'subject', 'value': 'Newspapers', 'r...   
4  [{'name': 'persons', 'value': 'Hollingsworth, ...   

                   pub_date  ... headline.kicker headline.content_kicker  \
0  2020-07-01T23:16:05+0000  ...            None                    None   
1  2020-07-01T14:06:12+0000  ...            None                    None   
2  2020-07-01T21:21:16+0000  ...            None                    None   
3  2020-07-03T02:13:01+0000  ...            None                    None   
4  2020-07-01T15:39:53+0000  ...            None                    None   

                             headline.print_headline headline.name  \
0  Afghan Contractor Gave Out Russian Cash, Offic...          None   
1  Blaming Gun Violence, Seattle Officials Clear ...          None   
2  Embassy Crisis in Riyadh Shows Perils of Diplo...          None   
3            Hedge Fund Delays Effort To Buy Tribune          None   
4   ‘You Can’t Just Get Up and Steal a Police Horse’          None   

  headline.seo headline.sub  \
0         None         None   
1         None         None   
2         None         None   
3         None         None   
4         None         None   

                                     byline.original  \
0  By Mujib Mashal, Eric Schmitt, Najim Rahim and...   
1                                   By Rachel Abrams   
2                   By Mark Mazzetti and Edward Wong   
3                                      By Marc Tracy   
4                                   By Ximena Larkin   

                                       byline.person byline.organization  \
0  [{'firstname': 'Mujib', 'middlename': None, 'l...                None   
1  [{'firstname': 'Rachel', 'middlename': None, '...                None   
2  [{'firstname': 'Mark', 'middlename': None, 'la...                None   
3  [{'firstname': 'Marc', 'middlename': None, 

In [7]:
data['pub_date'] = pd.to_datetime(data['pub_date'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6553 entries, 0 to 6552
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype              
---  ------                   --------------  -----              
 0   abstract                 6553 non-null   object             
 1   web_url                  6553 non-null   object             
 2   snippet                  6553 non-null   object             
 3   lead_paragraph           6553 non-null   object             
 4   print_section            4078 non-null   object             
 5   print_page               4078 non-null   object             
 6   source                   6553 non-null   object             
 7   multimedia               6553 non-null   object             
 8   keywords                 6553 non-null   object             
 9   pub_date                 6553 non-null   datetime64[ns, UTC]
 10  document_type            6553 non-null   object             
 11  news_desk                6553 

In [8]:
pd.json_normalize(data.iloc[0]['keywords'])

name                                      value  rank major
0      glocations                                     Russia     1     N
1   organizations                                    Taliban     2     N
2         persons                         Azizi, Rahmatullah     3     N
3      glocations                                AFGHANISTAN     4     N
4         subject                   Afghanistan War (2001- )     5     N
5         subject  United States Defense and Military Forces     6     N
6   organizations                               GRU (Russia)     7     N
7         subject        Espionage and Intelligence Services     8     N
8         persons                            Trump, Donald J     9     N
9         subject      United States International Relations    10     N
10        subject      United States Politics and Government    11     N

In [9]:
def entity_present(keywords: List[Dict[str, str]], entity: str) -> int:
    temp = pd.json_normalize(keywords)
    if 'value' in temp.columns:
        return temp['value'].isin([entity]).sum()
    else:
        return 0

def biden_present(keywords: List[Dict[str, str]]) -> int:
    return entity_present(keywords=keywords, entity='Biden, Joseph R Jr')
    
def trump_present(keywords: List[Dict[str, str]]) -> int:
    return entity_present(keywords=keywords, entity='Trump, Donald J')

In [10]:
data['biden_present'] = data.apply(lambda row: biden_present(keywords=row['keywords']), axis=1)
data['trump_present'] = data.apply(lambda row: trump_present(keywords=row['keywords']), axis=1)

data['both_present'] = data['biden_present'] * data['trump_present']
data['only_biden_present'] = data['biden_present'] * (1-data['both_present'])
data['only_trump_present'] = data['trump_present'] * (1-data['both_present'])
data.head()

abstract  \
0  A small-time businessman became a key middlema...   
1  The so-called Capitol Hill Organized Protest a...   
2  Surging outbreaks in the U.S. Embassy and the ...   
3  Journalists have been wary of Alden Global Cap...   
4  Adam Hollingsworth, known in Chicago as the Dr...   

                                             web_url  \
0  https://www.nytimes.com/2020/07/01/world/asia/...   
1  https://www.nytimes.com/2020/07/01/us/seattle-...   
2  https://www.nytimes.com/2020/07/01/us/politics...   
3  https://www.nytimes.com/2020/07/02/business/me...   
4  https://www.nytimes.com/2020/07/01/style/dread...   

                                             snippet  \
0  A small-time businessman became a key middlema...   
1  The so-called Capitol Hill Organized Protest a...   
2  Surging outbreaks in the U.S. Embassy and the ...   
3  Journalists have been wary of Alden Global Cap...   
4  Adam Hollingsworth, known in Chicago as the Dr...   

                                      lead_paragraph print_section print_page  \
0  KABUL, Afghanistan — He was a lowly drug smugg...             A         18   
1  SEATTLE — For weeks, officials in Seattle have...             A         20   
2  WASHINGTON — Inside the sprawling American Emb...             A          1   
3  Alden Global Capital seemed in position this w...             B          1   
4  In late May, as protests against police brutal...            ST          2   

               source                                         multimedia  \
0  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
1  The New York Times                                                 []   
2  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
3  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
4  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   

                                            keywords  \
0  [{'name': 'glocations', 'value': 'Russia', 'ra...   
1  [{'name': 'subject', 'value': 'George Floyd Pr...   
2  [{'name': 'subject', 'value': 'Coronavirus (20...   
3  [{'name': 'subject', 'value': 'Newspapers', 'r...   
4  [{'name': 'persons', 'value': 'Hollingsworth, ...   

                   pub_date  ... headline.sub  \
0 2020-07-01 23:16:05+00:00  ...         None   
1 2020-07-01 14:06:12+00:00  ...         None   
2 2020-07-01 21:21:16+00:00  ...         None   
3 2020-07-03 02:13:01+00:00  ...         None   
4 2020-07-01 15:39:53+00:00  ...         None   

                                     byline.original  \
0  By Mujib Mashal, Eric Schmitt, Najim Rahim and...   
1                                   By Rachel Abrams   
2                   By Mark Mazzetti and Edward Wong   
3                                      By Marc Tracy   
4                                   By Ximena Larkin   

                                       byline.person byline.organization  \
0  [{'firstname': 'Mujib', 'middlename': None, 'l...                None   
1  [{'firstname': 'Rachel', 'middlename': None, '...                None   
2  [{'firstname': 'Mark', 'middlename': None, 'la...                None   
3  [{'firstname': 'Marc', 'middlename': None, 'la...                None   
4  [{'firstname': 'Ximena', 'middlename': None, '...                None   

  slideshow_credits biden_present  trump_present both_present  \
0               NaN             0              1            0   
1               NaN             0              0            0   
2               NaN             0              0            0   
3               NaN             0              0            0   
4               NaN             0              0            0   

  only_biden_present only_trump_present  
0                  0                  1  
1                  0                  0  
2                  0                  0  
3                  0                  0  
4                  0                  0  

[5 rows x 34 column

In [11]:
data = data[(data['biden_present'] == 1) | (data['trump_present'] == 1)].reset_index(drop=True)
print(data.shape)
data.head()

(915, 34)


abstract  \
0  A small-time businessman became a key middlema...   
1  A reader offers an idea to spice up the Democr...   
2  The decision reversed a lower court’s ruling t...   
3  The experience of Representative Elissa Slotki...   
4  There’s a persuasive argument that the 2020 el...   

                                             web_url  \
0  https://www.nytimes.com/2020/07/01/world/asia/...   
1  https://www.nytimes.com/2020/07/02/opinion/let...   
2  https://www.nytimes.com/2020/07/01/us/politics...   
3  https://www.nytimes.com/2020/07/01/us/politics...   
4  https://www.nytimes.com/2020/07/01/opinion/tru...   

                                             snippet  \
0  A small-time businessman became a key middlema...   
1  A reader offers an idea to spice up the Democr...   
2  The decision reversed a lower court’s ruling t...   
3  The experience of Representative Elissa Slotki...   
4  There’s a persuasive argument that the 2020 el...   

                                      lead_paragraph print_section print_page  \
0  KABUL, Afghanistan — He was a lowly drug smugg...             A         18   
1                                     To the Editor:             A         22   
2  A New York appellate judge ruled on Wednesday ...             A         21   
3  WASHINGTON — From the moment President Trump p...             A         18   
4  Only two of the past six presidents before Don...           NaN        NaN   

               source                                         multimedia  \
0  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
1  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
2  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
3  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   
4  The New York Times  [{'rank': 0, 'subtype': 'xlarge', 'caption': N...   

                                            keywords  \
0  [{'name': 'glocations', 'value': 'Russia', 'ra...   
1  [{'name': 'subject', 'value': 'Presidential El...   
2  [{'name': 'creative_works', 'value': 'Too Much...   
3  [{'name': 'subject', 'value': 'United States D...   
4  [{'name': 'subject', 'value': 'Presidential El...   

                   pub_date  ... headline.sub  \
0 2020-07-01 23:16:05+00:00  ...         None   
1 2020-07-02 17:00:22+00:00  ...         None   
2 2020-07-01 22:56:01+00:00  ...         None   
3 2020-07-01 22:56:21+00:00  ...         None   
4 2020-07-01 09:00:20+00:00  ...         None   

                                     byline.original  \
0  By Mujib Mashal, Eric Schmitt, Najim Rahim and...   
1                                               None   
2                  By Maggie Haberman and Alan Feuer   
3                                  By Emily Cochrane   
4                                     By Frank Bruni   

                                       byline.person byline.organization  \
0  [{'firstname': 'Mujib', 'middlename': None, 'l...                None   
1                                                 []                None   
2  [{'firstname': 'Maggie', 'middlename': None, '...                None   
3  [{'firstname': 'Emily', 'middlename': None, 'l...                None   
4  [{'firstname': 'Frank', 'middlename': None, 'l...                None   

  slideshow_credits biden_present  trump_present both_present  \
0               NaN             0              1            0   
1               NaN             1              0            0   
2               NaN             0              1            0   
3               NaN             0              1            0   
4               NaN             1              1            1   

  only_biden_present only_trump_present  
0                  0                  1  
1                  1                  0  
2                  0                  1  
3                  0                  1  
4                  0                  0  

[5 rows x 34 column

# Sentiment Analysis

In [12]:
column_for_analysis = 'headline.main'
# column_for_analysis = 'abstract'

In [13]:
sentiment = pipeline("sentiment-analysis")

In [14]:
results = sentiment(data[column_for_analysis].tolist())

In [15]:
results = pd.json_normalize(results)

In [16]:
dfs = [data[['pub_date', column_for_analysis, 'biden_present', 'trump_present', 'only_biden_present', 'only_trump_present', 'both_present']], results]
sentiment = pd.concat(dfs, axis=1)
sentiment.head()

pub_date  \
0 2020-07-01 23:16:05+00:00   
1 2020-07-02 17:00:22+00:00   
2 2020-07-01 22:56:01+00:00   
3 2020-07-01 22:56:21+00:00   
4 2020-07-01 09:00:20+00:00   

                                       headline.main  biden_present  \
0  Afghan Contractor Handed Out Russian Cash to K...              0   
1         Let the Delegates Help Choose Biden’s V.P.              1   
2  Tell-All Book on Trump Can Move Forward Pendin...              0   
3  Slotkin, Former Intelligence Briefer, Presses ...              0   
4                                    Is Trump Toast?              1   

   trump_present  only_biden_present  only_trump_present  both_present  \
0              1                   0                   1             0   
1              0                   1                   0             0   
2              1                   0                   1             0   
3              1                   0                   1             0   
4              1                   0                   0             1   

      label     score  
0  NEGATIVE  0.987241  
1  POSITIVE  0.966687  
2  POSITIVE  0.944482  
3  NEGATIVE  0.908681  
4  POSITIVE  0.954712

In [17]:
def add_sign_to_score(row: pd.Series) -> float:
    if row['label'] == 'NEGATIVE':
        return row['score'] * -1
    else:
        return row['score']

sentiment['score'] = sentiment.apply(lambda row: add_sign_to_score(row=row), axis=1)
sentiment.head()

pub_date  \
0 2020-07-01 23:16:05+00:00   
1 2020-07-02 17:00:22+00:00   
2 2020-07-01 22:56:01+00:00   
3 2020-07-01 22:56:21+00:00   
4 2020-07-01 09:00:20+00:00   

                                       headline.main  biden_present  \
0  Afghan Contractor Handed Out Russian Cash to K...              0   
1         Let the Delegates Help Choose Biden’s V.P.              1   
2  Tell-All Book on Trump Can Move Forward Pendin...              0   
3  Slotkin, Former Intelligence Briefer, Presses ...              0   
4                                    Is Trump Toast?              1   

   trump_present  only_biden_present  only_trump_present  both_present  \
0              1                   0                   1             0   
1              0                   1                   0             0   
2              1                   0                   1             0   
3              1                   0                   1             0   
4              1                   0                   0             1   

      label     score  
0  NEGATIVE -0.987241  
1  POSITIVE  0.966687  
2  POSITIVE  0.944482  
3  NEGATIVE -0.908681  
4  POSITIVE  0.954712

In [18]:
print(sentiment.groupby('only_biden_present')['score'].agg(['mean', 'count']))
print(sentiment.groupby('only_trump_present')['score'].agg(['mean', 'count']))
print(sentiment.groupby('both_present')['score'].agg(['mean', 'count']))

mean  count
only_biden_present                 
0                  -0.464167    853
1                   0.158128     62
                        mean  count
only_trump_present                 
0                  -0.252974    205
1                  -0.470804    710
                  mean  count
both_present                 
0            -0.420294    772
1            -0.431214    143


# Conclusion

New York Times writes more articles with Trump as the subject than Biden (710+143 vs. 62+143). In general, when Trump is subject in in the article and not Biden, the article tends to have a negative sentiment (score: -0.47). When only Biden is the subject and not Trump, the sentimentiment is slightly positive (score: 0.15). When both are present as subjects, the tone is generally negative (score = -0.43).

# Exploration

In [19]:
sentiment.query("only_biden_present == 1").head()

pub_date                               headline.main  \
1  2020-07-02 17:00:22+00:00  Let the Delegates Help Choose Biden’s V.P.   
5  2020-07-03 09:00:24+00:00    Symone Sanders Bet on Biden, and Herself   
28 2020-07-01 15:41:47+00:00                Your Democratic ‘Dream Team’   
37 2020-07-02 23:25:03+00:00                      The View From Waukesha   
60 2020-07-03 09:00:24+00:00    Symone Sanders Bet on Biden, and Herself   

    biden_present  trump_present  only_biden_present  only_trump_present  \
1               1              0                   1                   0   
5               1              0                   1                   0   
28              1              0                   1                   0   
37              1              0                   1                   0   
60              1              0                   1                   0   

    both_present     label     score  
1              0  POSITIVE  0.966687  
5              0  POSITIVE  0.821591  
28             0  POSITIVE  0.998486  
37             0  POSITIVE  0.999181  
60             0  POSITIVE  0.821591

In [20]:
sentiment.query("only_trump_present == 1").head()
# You can see that the sentiment classifier is not perfect. The "Bounty' story is being classified as "POSITIVE". If we had more information about the articles, it might be helpful.

pub_date  \
0 2020-07-01 23:16:05+00:00   
2 2020-07-01 22:56:01+00:00   
3 2020-07-01 22:56:21+00:00   
7 2020-07-03 23:00:29+00:00   
8 2020-07-02 23:38:33+00:00   

                                       headline.main  biden_present  \
0  Afghan Contractor Handed Out Russian Cash to K...              0   
2  Tell-All Book on Trump Can Move Forward Pendin...              0   
3  Slotkin, Former Intelligence Briefer, Presses ...              0   
7              Reading Orwell for the Fourth of July              0   
8  Bounties Uproar Casts a Shadow Over a Rare Tru...              0   

   trump_present  only_biden_present  only_trump_present  both_present  \
0              1                   0                   1             0   
2              1                   0                   1             0   
3              1                   0                   1             0   
7              1                   0                   1             0   
8              1                   0                   1             0   

      label     score  
0  NEGATIVE -0.987241  
2  POSITIVE  0.944482  
3  NEGATIVE -0.908681  
7  POSITIVE  0.946267  
8  POSITIVE  0.997181

# Backup

## Named Entity Recognition

### Conclusion: It looks like the NER provided by NYTimes is better since they have access to the full atricle.

In [21]:
ner = pipeline("ner")

In [22]:
results = ner(data[column_for_analysis].tolist())

KeyboardInterrupt: 

In [22]:
print(data.iloc[8][column_for_analysis])
print(pd.json_normalize(results[8]))
print(pd.json_normalize(data.iloc[7]['keywords']))
print(sentiment.iloc[8])

Bounties Uproar Casts a Shadow Over a Rare Trump Foreign Policy Achievement
    word     score  entity  index
0  Trump  0.727731  I-MISC     14
             name                       value  rank major
0         persons              Orwell, George     1     N
1         persons             Trump, Donald J     2     N
2         persons                  Xi Jinping     3     N
3         persons           Putin, Vladimir V     4     N
4   organizations                Facebook Inc     5     N
5         subject                Social Media     6     N
6         subject        Freedom of the Press     7     N
7      glocations                       China     8     N
8      glocations                   Hong Kong     9     N
9      glocations                      Poland    10     N
10     glocations                      Russia    11     N
11     glocations               United States    12     N
12     glocations  USSR (Former Soviet Union)    13     N
pub_date                                    